### PR TITLE
Printing

### PR DESCRIPTION
**Pull request by twiss**
WIP, because mobile UX could be better. In particular, if someone wants to test Chrome iOS that would be much appreciated.

### CODE REVIEW

**Comment by ferndot** [comment]
Works well in Firefox Desktop.

You might want to check if `window.print()` is available, and hide the menu item on platforms that do not support it, such as Firefox OS.

**Comment by twiss** [comment]
> hide the menu item on platforms that do not support it, such as Firefox OS

I would like to do this, but

> You might want to check if window.print() is available

have you tested this? Because window.print is there on Firefox and Chrome android, it just doesn't do anything (https://crbug.com/322303). Also, we probably can't tell if printing is available at all (https://addons.mozilla.org/mobile/addon/cloud-printer/). So maybe we should rename the button to "Print view" or something on non-desktop?

**Comment by ferndot** [comment]
What about: http://stackoverflow.com/a/9273097

**Comment by twiss** [comment]
Well, you have to call print first before you can know whether it worked. So we could point out that they have to select print from their browser menu, but we don't even know if it's there.

**Comment by ferndot** [comment]
> Well, you have to call print first before you can know whether it worked.

My idea is that we use the current implementation, and then alert the user and close the popup if the print fails (instead of doing nothing).

**Comment by twiss** [comment]
But there are also browsers where `window.print()` doesn't work, but there is in fact an option to print in the browser menu.

**Comment by ferndot** [comment]
Hmm, maybe we could include that in the alert. It could say, "We cannot print this file automatically. Please select 'Print' from inside your browser if it is available."

**Comment by twiss** [comment]
Done. Could you add both l10n ids? Also, this probably requires some more thinking about security than I did (postMessage, untrusted documents).

**Comment by ferndot** [comment]
I will add the IDs. What security concerns did you have?

File: scripts/firetext.js [hunk 1]
```diff
@@ -1061,6 +1061,106 @@ function processActions(eventAttribute, target) {
...
+				var reader = new FileReader();
+				reader.addEventListener('load', function() {
+					win.postMessage({
+						content: reader.result,
+						type: e.data.type
+					}, '*');
+				});
+				reader.readAsText(new Blob([
+					[
+						'<!DOCTYPE html>',
```
Comment (ferndot) [inline scripts/firetext.js]: This does not block merging, but is there a way that we could put this in an HTML file?

Comment (twiss) [inline scripts/firetext.js]: It's possible, but you would probably need similar tricks to those in #246 to get it working in Firefox (not that that's bad, maybe we could unify the document displaying of both places).

**Comment by twiss** [comment]
> What security concerns did you have?

As I said, I haven't thought about it at all, but basically:
- MDN always warns about postMessage with target `"*"`, and we're doing just that
- Documents aren't sandboxed

**Comment by ferndot** [comment]
> MDN always warns about postMessage with target "*", and we're doing just that

IIRC @jackd1 did some sort of workaround for this in the module loader. Jack, could you clarify?

**Comment by jackd1** [comment]
I used XHR to get the html, then replaced all occurrences of [ORIGIN_OF_MAIN_DOCUMENT] in all of the scripts with the actual origin, basically like an include of sorts.  Failing to replace it would result in the script not being authenticated aND refusing communication with the parent.

**Comment by twiss** [comment]
There's another call to postMessage \* in this PR though, although it doesn't seem like a problem to me.

That leaves sandboxing. There is CSP sandbox, which is identical to iframe sandbox, but Firefox doesn't support it (https://bugzil.la/671389). Another possibility is maybe somehow getting the window to be a different origin than the opener, which (I think) happens if the opener doesn't have an origin, so that could work.

**Comment by twiss** [comment]
> Another possibility is maybe somehow getting the window to be a different origin than the opener, which (I think) happens if the opener doesn't have an origin, so that could work.

I tried this and not surprisingly it doesn't work, probably for the same https://bugzil.la/1095808 mentioned in the bug for this (https://github.com/codexa/firetext/issues/206#issuecomment-62205811).

My only idea left is manually sanitizing the document, I'll file a bug for that.

**Comment by twiss** [comment]
I've changed my mind, among other reasons because https://bugzil.la/1095808 seems to be specific to printing to a file.

The document to be printed is now sandboxed, for details see commit message.

**Comment by ferndot** [comment]
That looks like a good solution :+1:

**Pull request merged**